syntax = "proto3";

import "google/protobuf/empty.proto";
package controller;

enum MessageType {
  UNSPECIFIED = 0; // unknown command type
  ACK = 1; // acknowledged
  APPEND_FUNCTION = 2; // append function
  APPEND_FUNCTION_ARG = 3; // append function args
  APPEND_CLASS = 4; // append class
  APPEND_CLASS_METHOD_ARG = 5; // append class method
  INVOKE_FUNCTION = 6; // invoke function
  INVOKE_CLASS_METHOD = 7; // invoke class method
  RT_RESULT = 8; // return result
  RT_CLASS_METHOD_RESULT = 9; // return class method result
}

message Data {
    enum ObjectType {
        OBJ_UNSPECIFIED = 0;
        OBJ_REF = 1;
        OBJ_ENCODED = 2;
        OBJ_STREAM = 3;
    }
    
    ObjectType type = 1;
    oneof Object {
        string ref = 2;
        bytes encoded = 3;
    }
    int64 size = 4; // size in bytes
}

message Ack {
    string message = 1; // ack message
    string error = 2;
}

message AppendFunction {
    string function_name = 1; // function name
    bytes function_code = 2; // function code
    repeated string params = 3; // function params
    repeated string requirements = 4; // function dependencies
}

message AppendFunctionArg {
    string session_id = 1; // current execution session
    string instance_id = 2; // id of function instance under current execution session
    string function_name = 3; // function name
    string param_name = 4; // param name
    Data value = 5; // object or ref for param
}

message AppendClass {
    message Method {
        string method_name = 1; // method name
        repeated string params = 2; // method params
    }
    string class_name = 1; // class name
    bytes class_code = 2; // class code
    repeated Method methods = 3; // class methods
}

message AppendClassMethodArg {
    string session_id = 1; // current execution session
    string instance_id = 2; // id of class instance under current execution session
    string function_id = 3; // id of function instance under current execution session
    string class_name = 4; // class name
    string method_name = 5; // method name
    string param_name = 6; // param name
    Data value = 7; // object or ref for param
}

message InvokeFunction {
    string session_id = 1; // current execution session
    string instance_id = 2; // id of function instance under current execution session
    string function_name = 3; // function name
}

message ReturnResult {
    string session_id = 1; // current execution session
    string instance_id = 2; // id of function instance under current execution session
    string function_name = 3; // function name
    oneof result {
        Data value = 4; // success: return result
        string error = 5; // fail: return error
    }
}

message InvokeClassMethod {
    string session_id = 1; // current execution session
    string instance_id = 2; // id of class instance under current execution session
    string function_id = 3; // id of function instance under current execution session
    string class_name = 4; // class name
    string method_name = 5; // method name
}

message ReturnClassMethodResult {
    string session_id = 1; // current execution session
    string instance_id = 2; // id of class instance under current execution session
    string function_id = 3; // id of function instance under current execution session
    string class_name = 4; // class name
    string method_name = 5; // method name
    oneof result {
        Data value = 6; // success: return result
        string error = 7; // fail: return error
    }
}

message Message {
    MessageType type = 1;
    oneof Message {
        Ack ack = 2;
        AppendFunction append_function = 3;
        AppendFunctionArg append_function_arg = 4;
        AppendClass append_class = 5;
        AppendClassMethodArg append_class_method_arg = 6;
        InvokeFunction invoke_function = 7;
        InvokeClassMethod invoke_class_method = 8;
        ReturnResult return_result = 9;
        ReturnClassMethodResult return_class_method_result = 10;
    }
}

service ControllerService {
    rpc Session(stream Message) returns (stream Message) {}
    rpc Ping(google.protobuf.Empty) returns (Ack) {}
}